src/wrappers/pubsub.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * @fileoverview Handlers for Pubsub instrumentation
+ */
+
+const shimmer = require('shimmer');
+const {
+    tracer,
+    tryRequire,
+    eventInterface,
+} = require('epsagon');
+const traceContext = require('../trace_context.js');
+
+const subscriber = tryRequire('@google-cloud/pubsub/build/src/subscriber');
+
+/**
+ * Handle subscriber event emitter of eventName='message'
+ * @param {Message} message received message.
+ * @param {*} originalHandler listener callback function.
+ * @param {*} requestFunctionThis request arguments.
+ */
+function pubSubSubscriberMiddleware(message, originalHandler, requestFunctionThis) {
+    let originalHandlerErr;
+    try {
+        // Initialize tracer and evnets.
+        tracer.restart();
+        const functionName = originalHandler.name || 'messageHandler';
+        const { slsEvent: pubSubEvent, startTime: pubSubStartTime } =
+        eventInterface.initializeEvent(
+            'pubsub',
+            requestFunctionThis.projectId,
+            'messagePullingListener',
+            'trigger'
+        );
+        const { slsEvent: nodeEvent, startTime: nodeStartTime } = eventInterface.initializeEvent(
+            'node_function', functionName, 'messageReceived', 'runner'
+        );
+        tracer.addEvent(pubSubEvent);
+
+        // Getting message data.
+        let callbackResponse = { messageId: message.id };
+        const messageData = (message.data && JSON.parse(`${message.data}`));
+        if (messageData && typeof messageData === 'object') {
+            callbackResponse = Object.assign(callbackResponse, messageData);
+        }
+        const { label, setError } = tracer;
+        const returnMessage = message;
+        returnMessage.epsagon = {
+            label,
+            setError,
+        };
+
+        // Finalize pubsub event.
+        eventInterface.finalizeEvent(pubSubEvent, pubSubStartTime, null, callbackResponse);
+        let promise;
+        try {
+            promise = originalHandler(returnMessage, {});
+        } catch (err) {
+            originalHandlerErr = err;
+        }
+        // Handle and finalize async user function.
+        if (promise && promise.then) {
+            promise.catch((err) => {
+                originalHandlerErr = err;
+                throw err;
+            }).finally(() => {
+                eventInterface.finalizeEvent(nodeEvent, nodeStartTime, originalHandlerErr);
+                tracer.sendTrace(() => {});
+            });
+        } else {
+            // Finalize sync user function.
+            eventInterface.finalizeEvent(nodeEvent, nodeStartTime, originalHandlerErr);
+            tracer.sendTrace(() => {});
+        }
+        tracer.addRunner(nodeEvent, promise);
```

Comment:
if the user returns '3' (and not a promise), your code will explode.

Reply:
you can add a better validation in `tracer.addRunner` instead of fixing it here

Reply:
Added validation to tracer.addRunner.